File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:

int calls=0;
    void funcOptimal(int ind, int k, vector<int>& arr, vector<vector<int>>& ans,
                     vector<int>& ds) {

        if (k == 0) {
            ans.push_back(ds);
            return;
        }
        calls++;

        if(ind==arr.size()) return ;

        for (int i = ind; i < arr.size(); i++) {
            if (i > ind && arr[i] == arr[i - 1])
                continue;
            if (arr[i] > k)
                break;
            ds.push_back(arr[i]);
            funcOptimal(i + 1, k - arr[i], arr, ans, ds);
            ds.pop_back();
        }
    }
    vector<vector<int>> combinationSum2(vector<int>& arr, int k) {

        vector<int> ds;
        vector<vector<int>> ans;
        sort(arr.begin(), arr.end());
        funcOptimal(0, k, arr, ans, ds);
    
        return ans;
    }
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    
    public void func(int i, int[] candidates, int target,List<List<Integer>> ans , List<Integer> ds, int n){
        if(target==0){
            // vv imp , cannot simply add ds 
            ans.add(new ArrayList<>(ds));
        }
        if(i>=n|| target<0){
            return;
        }
        for (int j = i; j < n; j++) {
            if (j > i && candidates[j] == candidates[j-1]) continue; // skip duplicate
            ds.add(candidates[j]);
            func(j+1, candidates, target - candidates[j], ans, ds, n);
            ds.remove(ds.size()-1);
        }

    }
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> ans=new ArrayList<>();
        List<Integer> ds=new ArrayList<>();
        Arrays.sort(candidates);
        func(0,candidates, target, ans, ds, candidates.length);
        return ans;

    }
}